### PR TITLE
Region Proposal Network (RPN) corrections

### PR DESCRIPTION
**Pull request by hgaiser**
This is basically parts copied from https://github.com/broadinstitute/keras-rcnn/pull/51 that are truly necessary. https://github.com/broadinstitute/keras-rcnn/pull/51 was quite huge and only intended to demonstrate how I think it should look like. This PR is intended to be merged into `features/region-proposal-network-losses`. It also rebased on the current master.

### CODE REVIEW

**Comment by hgaiser** [comment]
With the exception of the TODO's mentioned in the code, I think the `ObjectProposal` layer should work (in theory) with this PR.

Something I noticed for which I didn't have a fix yet, the number of maximum NMS proposals differs depending on whether we are [training](https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/fast_rcnn/config.py#L114) or [testing](https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/fast_rcnn/config.py#L162). Currently in `keras-rcnn` the value is set to 300, always.

A bigger question is hiding in there, which is what do we do about configuration such as this?

**Comment by jhung0** [comment]
Ready to merge?
Yes, currently we have much fewer config options than the original  implementation. We could create a class (like in https://github.com/yhenon/keras-frcnn/blob/master/keras_frcnn/config.py), but I guess that's not ideal @0x00b1 ?

**Comment by hgaiser** [comment]
@jhunh0 yeah I'm done :) if you agree with the changes I've made then it is good to merge.

I agree, a decision should be made on how to handle configuration.

File: keras_rcnn/layers/object_detection/proposal_target.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        labels = ones * -1
+
+        argmax_overlaps_inds, max_overlaps, gt_argmax_overlaps_inds = keras_rcnn.backend.overlapping(anchors, gt_boxes, indices)
+
+        if not self.clobber_positives:
+            labels = keras_rcnn.backend.where(keras.backend.less(max_overlaps, self.negative_overlap), zeros, labels)
+
+        # fg label: for each gt, anchor with highest overlap
+        #TODO Find Keras equivalent of this
+        #labels[gt_argmax_overlaps_inds] = 1
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/proposal_target.py]: You could use something like:

`keras_rcnn.backend.scatter_add_tensor(labels, indices, keras.backend.ones((...)))`

File: keras_rcnn/layers/object_detection/proposal_target.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import tensorflow
+
+import keras_rcnn.backend
+
+
+class ProposalTarget(keras.engine.topology.Layer):
+    def __init__(self, **kwargs):
+        self.allowed_border = 0
+        self.clobber_positives = False
+        self.negative_overlap = 0.3
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/proposal_target.py]: I would consider adding this as an argument to `__init__`.

File: keras_rcnn/layers/object_detection/proposal_target.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.positive_overlap = 0.7
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/proposal_target.py]: I would consider adding this as an argument to `__init__`.

File: keras_rcnn/layers/object_detection/proposal_target.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        #TODO Find Keras equivalent of this
+        #labels[gt_argmax_overlaps_inds] = 1
+
+        # fg label: above threshold IOU
+        labels = keras_rcnn.backend.where(keras.backend.greater_equal(max_overlaps, self.positive_overlap), ones, labels)
+
+        if self.clobber_positives:
+            labels = keras_rcnn.backend.where(keras.backend.less(max_overlaps, self.negative_overlap), zeros, labels)
+
+        # subsample positive labels if we have too many
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/proposal_target.py]: Have you tried using the `balance` function? It should subsample positive and negative labels.

File: keras_rcnn/layers/object_detection/object_proposal.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import keras.engine
+import tensorflow
+
+import keras_rcnn.backend
+
+
+class ObjectProposal(keras.engine.topology.Layer):
+    def __init__(self, maximum_proposals=300, **kwargs):
+        # TODO : Parametrize this
+        self.min_size    = 16 # minimum width/height of proposals in original image size
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/object_proposal.py]: I would consider adding this as an argument to `__init__`.

File: keras_rcnn/layers/object_detection/object_proposal.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        self.feat_stride = 16
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/object_proposal.py]: I would consider adding this as an argument to `__init__`.

File: keras_rcnn/layers/object_detection/object_proposal.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import keras.backend
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/object_proposal.py]: The module should be `_object_proposal.py` rather than `object_proposal.py`.

This will fix the current Travis failure.

File: keras_rcnn/layers/object_detection/proposal_target.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+        # label: 1 is positive, 0 is negative, -1 is dont care
+        ones   = keras.backend.ones_like(indices, dtype=keras.backend.floatx())
+        zeros  = keras.backend.zeros_like(indices, dtype=keras.backend.floatx())
+        labels = ones * -1
+
+        argmax_overlaps_inds, max_overlaps, gt_argmax_overlaps_inds = keras_rcnn.backend.overlapping(anchors, gt_boxes, indices)
+
+        if not self.clobber_positives:
+            labels = keras_rcnn.backend.where(keras.backend.less(max_overlaps, self.negative_overlap), zeros, labels)
```
Comment (jhung0) [inline keras_rcnn/layers/object_detection/proposal_target.py]: is `where` in `keras_rcnn.backend`?

File: keras_rcnn/layers/object_detection/object_proposal.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        indices = keras_rcnn.backend.filter_boxes(proposals, self.min_size * scale)
+        proposals = keras.backend.gather(proposals, indices)
+        scores = scores[..., (scores.shape[-1] // 2):]
+        scores = keras.backend.reshape(scores, (-1, 1))
+        scores = keras.backend.gather(scores, indices)
+        scores = keras.backend.flatten(scores)
+
+        # 4. sort all (proposal, score) pairs by score from highest to lowest
+        # 5. take top pre_nms_topN (e.g. 6000)
+        # TODO : Needs to be implemented?
```
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/object_proposal.py]: Yep. Would you like some advice? Sorting is kind of a nightmare.

File: keras_rcnn/layers/object_detection/proposal_target.py (hunk 2 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/proposal_target.py]: Yeah that'd definitely be a step up. I figured lets leave it hardcoded for now until we have a place to configure everything nicely. Regardless, `__init__` arguments are better.

File: keras_rcnn/layers/object_detection/proposal_target.py (hunk 3 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/proposal_target.py]: See comment above.

File: keras_rcnn/layers/object_detection/proposal_target.py (hunk 1 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/proposal_target.py]: I tried the same thing, but `scatter` requires that its input is a `tf.Variable`, but `labels` is a `tf.Tensor`. Any experience with this?

File: keras_rcnn/layers/object_detection/proposal_target.py (hunk 4 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/proposal_target.py]: No haven't looked at that yet, I'll check it out.

File: keras_rcnn/layers/object_detection/object_proposal.py (hunk 5 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/object_proposal.py]: Same.

File: keras_rcnn/layers/object_detection/object_proposal.py (hunk 6 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/object_proposal.py]: Same.

File: keras_rcnn/layers/object_detection/object_proposal.py (hunk 7 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/object_proposal.py]: I was wondering what the reason was for the underscore in the name, none of the other modules have this naming convention?

File: keras_rcnn/layers/object_detection/object_proposal.py (hunk 9 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/object_proposal.py]: Be my guest! :D I'm not sure when I can continue work on this, got some other stuff to work on.

File: keras_rcnn/layers/object_detection/proposal_target.py (hunk 1 above)
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/proposal_target.py]: Yep. The `keras_rcnn.backend.scatter_add_tensor` will deal with this problem.

File: keras_rcnn/layers/object_detection/object_proposal.py (hunk 7 above)
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/object_proposal.py]: It provides a private namespace so we can keep modules short and simple. Users would import `ObjectProposal` from `keras_rcnn.layers.object_detection`.

Comment (hgaiser) [inline keras_rcnn/layers/object_detection/object_proposal.py]: But the filename `_object_proposal.py` or `object_proposal.py` makes no difference in this scenario? As long as you import the class from the `object_detection` module.

File: keras_rcnn/layers/object_detection/proposal_target.py (hunk 8 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/proposal_target.py]: It was added in https://github.com/broadinstitute/keras-rcnn/pull/41

File: keras_rcnn/layers/object_detection/object_proposal.py (hunk 7 above)
Comment (0x00b1) [inline keras_rcnn/layers/object_detection/object_proposal.py]: Ah. I misunderstood. This is a style convention:

https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

Comment (hgaiser) [inline keras_rcnn/layers/object_detection/object_proposal.py]: Ah I see, I was not aware of this. I'll change this back.

**Comment by hgaiser** [comment]
What I think is still missing in this PR:

- I stopped porting `anchor_target_layer` until [here](https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/rpn/anchor_target_layer.py#L147). The rest of the `forward` method also needs to be ported.
- The losses need to be implemented as Keras layers (a la https://github.com/fchollet/keras/blob/master/examples/variational_autoencoder.py#L46-L62)

I believe these two things should be enough to attempt to train the RPN part of RCNN. For the head of RCNN, [`proposal_target_layer`](https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/rpn/proposal_target_layer.py) is also required (preferably as one layer?) and additional loss functions need to be made.

In addition it wouldn't hurt to have unit tests and documentation for all this :)
